hibernate-core/src/main/java/org/hibernate/bytecode/internal/bytebuddy/ByteBuddyState.java

```diff
@@ -81,8 +88,41 @@ public static ByteBuddy getStaticByteBuddyInstance() {
 		return buddy;
 	}
 
-	public static ClassLoadingStrategy getLoadingStrategy() {
-		return loadingStrategy;
+	public static ClassLoadingStrategy<?> resolveClassLoadingStrategy(Class<?> originalClass) {
+		if ( ClassInjector.UsingLookup.isAvailable() ) {
+			Object lookup;
+			Method privateLookupIn;
+			try {
+				Class<?> methodHandles = Class.forName( "java.lang.invoke.MethodHandles" );
+				lookup = methodHandles.getMethod( "lookup" ).invoke( null );
+				privateLookupIn = methodHandles.getMethod( "privateLookupIn", Class.class, Class.forName( "java.lang.invoke.MethodHandles$Lookup" ) );
+			}
+			catch (Exception e) {
+				throw new HibernateException( LOG.bytecodeEnhancementFailed( originalClass.getName() ), e );
+			}
+			Object privateLookup;
+			try {
+				try {
+					privateLookup = privateLookupIn.invoke( null, originalClass, lookup );
+				}
+				catch (InvocationTargetException exception) {
+					if ( exception.getCause() instanceof IllegalAccessException ) {
+						return new ClassLoadingStrategy.ForUnsafeInjection( originalClass.getProtectionDomain() );
+					}
+					else {
+						throw new HibernateException( LOG.bytecodeEnhancementFailed( originalClass.getName() ), exception.getCause() );
+					}
+				}
+
+				return ClassLoadingStrategy.UsingLookup.of( privateLookup );
+			}
+			catch (Throwable e) {
+				throw new HibernateException( LOG.bytecodeEnhancementFailedUnableToGetPrivateLookupFor( originalClass.getName() ), e );
```

Comment:
this new error message gives some very specific advice to users.

Yet it's catching `Throwable` .  How confident are you that this suggestion to open up the packages won't be misleading, e.g. some other error is actually happening?

I would either rephrase the error to mention that solution as one possible cause (among many) - or ideally be very specific on what we're catching.

Reply:
So as for the message rationale, I followed what Rafael did on Mockito. I agree we should probably be more cautious about the advice we give.